### DISCUSSION TITLE
How to enforce the templ cli version to match the templ module version?

### DISCUSSION THREAD

**Discussion by yardenshoham** [a-h/templ#592] [Q&A]
As far as I understand, templ is used in 2 ways in a project. The first is as a dependency of `github.com/a-h/templ` in `go.mod`, specifying the version in `go.mod` and thus making sure you can `git clone` then `go mod download` and you end up using the intended version of templ. This is needed as the generated code imports `github.com/a-h/templ`.

The second way is as a CLI tool, installed with `go install github.com/a-h/templ/cmd/templ@VERSION`. This is to allow running `templ generate`. I don't see a builtin way to enforce the versions of these to match. The best thing I can think of is `go install github.com/a-h/templ/cmd/templ@$(go list -m -f '{{ .Version }}' github.com/a-h/templ)` but this is a one-time operation, not suitable in case of a templ version upgrade. `go mod download` doesn't control the version of the templ CLI tool.

I am looking for a development experience similar to that of NodeJS tools like typescript or prettier. I can enforce the typescript compiler version by specifying the typescript module version in `package.json`, then everyone use `npx tsc` or `npm run build` and the version is maintained in `package.json`. With the templ CLI tool there is a different experience, there's no equivalent to `npm install` as `go mod download` doesn't handle the templ CLI tool version inference.

**Comment by joerdav** [comment]
One option for you would be to not use `go install` then `templ generate`.

I believe that you can use `go run` to use the version specified in the go.mod. 

So the command would just be:

`go run github.com/a-h/templ/cmd/templ generate`

I'm on my phone so haven't had chance to test it out, but I believe it is possible.

**Reply by yardenshoham** [reply]
I tried but it fails with

```
$ go run github.com/a-h/templ/cmd/templ generate
../go/pkg/mod/github.com/a-h/templ@v0.2.598/parser/v2/calltemplateparser.go:4:2: missing go.sum entry for module providing package github.com/a-h/parse (imported by github.com/a-h/templ/parser/v2); to add:
        go get github.com/a-h/templ/parser/v2@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/fmtcmd/main.go:16:2: missing go.sum entry for module providing package github.com/natefinch/atomic (imported by github.com/a-h/templ/cmd/templ/fmtcmd); to add:
        go get github.com/a-h/templ/cmd/templ/fmtcmd@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/generatecmd/cmd.go:23:2: missing go.sum entry for module providing package github.com/cenkalti/backoff/v4 (imported by github.com/a-h/templ/cmd/templ/generatecmd); to add:
        go get github.com/a-h/templ/cmd/templ/generatecmd@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/generatecmd/cmd.go:24:2: missing go.sum entry for module providing package github.com/cli/browser (imported by github.com/a-h/templ/cmd/templ/generatecmd); to add:
        go get github.com/a-h/templ/cmd/templ/generatecmd@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/lspcmd/main.go:11:2: missing go.sum entry for module providing package github.com/a-h/protocol (imported by github.com/a-h/templ/cmd/templ/lspcmd); to add:
        go get github.com/a-h/templ/cmd/templ/lspcmd@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/lspcmd/proxy/server.go:14:2: missing go.sum entry for module providing package go.lsp.dev/uri (imported by github.com/a-h/templ/cmd/templ/lspcmd/proxy); to add:
        go get github.com/a-h/templ/cmd/templ/lspcmd/proxy@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/cmd/templ/lspcmd/main.go:15:2: missing go.sum entry for module providing package go.lsp.dev/jsonrpc2 (imported by github.com/a-h/templ/cmd/templ/lspcmd); to add:
        go get github.com/a-h/templ/cmd/templ/lspcmd@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/parser/v1/elementparser.go:9:2: missing go.sum entry for module providing package github.com/a-h/lexical/input (imported by github.com/a-h/templ/parser/v1); to add:
        go get github.com/a-h/templ/parser/v1@v0.2.598
../go/pkg/mod/github.com/a-h/templ@v0.2.598/parser/v1/calltemplateparser.go:6:2: missing go.sum entry for module providing package github.com/a-h/lexical/parse (imported by github.com/a-h/templ/parser/v1); to add:
        go get github.com/a-h/templ/parser/v1@v0.2.598
```

Should I try running these `go get`s it suggests?

**Reply by silverwind** [reply]
Yes, that also what I [was told](https://github.com/golang/go/issues/65637#issuecomment-1998729534) that if you `go run github.com/a-h/templ/cmd/templ` it will use the version from go.mod, if it is imported anywhere in your project.

Long-term it's a problem of the go tooling and https://github.com/golang/go/issues/48429 would help.
